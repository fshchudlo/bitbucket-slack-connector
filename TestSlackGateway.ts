import * as slack from "@slack/web-api";
import { SlackGateway } from "./contracts";

export class TestSlackGateway implements SlackGateway {
    snapshot: {
        createdChannels: slack.ConversationsCreateArguments[];
        setChannelTopics: slack.ConversationsSetTopicArguments[];
        invitesToChannels: slack.ConversationsInviteArguments[];
        closedChannels: slack.ConversationsCloseArguments[];
        sentMessages: slack.ChatPostMessageArguments[];
        lookedUpUsers: slack.UsersLookupByEmailArguments[];
    };
    constructor() {
        this.snapshot = {
            createdChannels: new Array<slack.ConversationsCreateArguments>(),
            setChannelTopics: new Array<slack.ConversationsSetTopicArguments>(),
            invitesToChannels: new Array<slack.ConversationsInviteArguments>(),
            closedChannels: new Array<slack.ConversationsCloseArguments>(),
            sentMessages: new Array<slack.ChatPostMessageArguments>(),
            lookedUpUsers: new Array<slack.UsersLookupByEmailArguments>(),
        };
    }
    lookupUserByEmail(options: slack.UsersLookupByEmailArguments): Promise<slack.UsersLookupByEmailResponse> {
        this.snapshot.lookedUpUsers.push(options);
        return Promise.resolve({ ok: true, user: { id: options.email } });
    }
    createChannel(options: slack.ConversationsCreateArguments): Promise<slack.ConversationsCreateResponse> {
        this.snapshot.createdChannels.push(options);
        return Promise.resolve({ ok: true, channel: { id: "12345", name: options.name } });
    }
    setChannelTopic(options: slack.ConversationsSetTopicArguments): Promise<slack.ConversationsSetTopicResponse> {
        this.snapshot.setChannelTopics.push(options);
        return Promise.resolve({ ok: true });
    }
    inviteToChannel(options: slack.ConversationsInviteArguments): Promise<slack.ConversationsInviteResponse> {
        this.snapshot.invitesToChannels.push(options);
        return Promise.resolve({ ok: true });
    }
    closeChannel(options: slack.ConversationsCloseArguments): Promise<slack.ConversationsCloseResponse> {
        this.snapshot.closedChannels.push(options);
        return Promise.resolve({ ok: true });
    }
    sendMessage(options: slack.ChatPostMessageArguments): Promise<slack.ChatPostMessageResponse> {
        this.snapshot.sentMessages.push(options);
        return Promise.resolve({ ok: true });
    }
}
