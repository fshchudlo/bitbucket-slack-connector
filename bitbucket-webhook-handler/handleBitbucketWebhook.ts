import * as useCases from "./use-cases";
import { SlackAPIAdapter } from "./ports/SlackAPIAdapter";
import { BitbucketAPIAdapter } from "./ports/BitbucketAPIAdapter";
import { BitbucketNotification } from "../bitbucket-payload-types";
import { WebhookHandlerConfig } from "./webhookHandlerConfig";
import { provisionPullRequestChannel } from "./provisionPullRequestChannel";

export default async function handleBitbucketWebhook(payload: BitbucketNotification, slackAPI: SlackAPIAdapter, bitbucketAPI: BitbucketAPIAdapter, config: WebhookHandlerConfig) {
    const eventKey = payload.eventKey;
    const channelInfo = await provisionPullRequestChannel(slackAPI, bitbucketAPI, payload, config);

    switch (eventKey) {
        case "pr:opened":
            await useCases.inviteParticipantsAndSetChannelBookmark(payload, slackAPI, config.defaultChannelParticipants, channelInfo.id);
            await useCases.tryBroadcastMessageAboutOpenedPR(payload, slackAPI, config.getOpenedPRBroadcastChannelId(payload));
            break;
        case "pr:modified":
            await useCases.sendMessageAboutPRModification(payload, slackAPI, channelInfo.id);
            break;
        case "pr:reviewer:updated":
            await useCases.updateChannelMembers(payload, slackAPI, channelInfo.id);
            break;
        case "pr:reviewer:unapproved":
        case "pr:reviewer:needs_work":
        case "pr:reviewer:approved":
            await useCases.sendMessageAboutReviewerAction(payload, slackAPI, channelInfo.id);
            break;
        case "pr:comment:added":
            await useCases.sendMessageAboutAddedComment(payload, slackAPI, channelInfo.id);
            break;
        case "pr:comment:edited":
            await useCases.sendMessageAboutEditedComment(payload, slackAPI, channelInfo.id);
            break;
        case "pr:comment:deleted":
            await useCases.sendMessageAboutDeletedComment(payload, slackAPI, channelInfo.id);
            break;
        case "pr:from_ref_updated":
            await useCases.sendMessageAboutNewCommit(payload, slackAPI, bitbucketAPI, channelInfo.id);
            break;
        case "pr:merged":
        case "pr:declined":
        case "pr:deleted":
            await useCases.sendCompletionMessageAndCloseTheChannel(payload, slackAPI, channelInfo.id);
            await useCases.tryBroadcastMessageAboutClosedPR(payload, slackAPI, config.getOpenedPRBroadcastChannelId(payload));
            break;
        default:
            throw new Error(`"${eventKey}" event key is unknown.`);
    }
}

