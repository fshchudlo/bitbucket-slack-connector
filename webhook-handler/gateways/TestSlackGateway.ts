import * as slack from "@slack/web-api";

import { SlackGateway } from "./SlackGateway";

const channelId = "12345";
export default class TestSlackGateway implements SlackGateway {
    snapshot: {
        createdChannels: slack.ConversationsCreateArguments[];
        setChannelTopics: slack.ConversationsSetTopicArguments[];
        invitesToChannels: slack.ConversationsInviteArguments[];
        archivedChannels: slack.ConversationsCloseArguments[];
        sentMessages: slack.ChatPostMessageArguments[];
        lookedUpUsers: slack.UsersLookupByEmailArguments[];
    };

    constructor() {
        this.snapshot = {
            createdChannels: new Array<slack.ConversationsCreateArguments>(),
            setChannelTopics: new Array<slack.ConversationsSetTopicArguments>(),
            invitesToChannels: new Array<slack.ConversationsInviteArguments>(),
            archivedChannels: new Array<slack.ConversationsCloseArguments>(),
            sentMessages: new Array<slack.ChatPostMessageArguments>(),
            lookedUpUsers: new Array<slack.UsersLookupByEmailArguments>()
        };
    }

    lookupUserByEmail(options: slack.UsersLookupByEmailArguments): Promise<slack.UsersLookupByEmailResponse> {
        this.snapshot.lookedUpUsers.push(options);
        return Promise.resolve({ ok: true, user: { id: options.email } });
    }

    createChannel(options: slack.ConversationsCreateArguments): Promise<slack.ConversationsCreateResponse> {
        this.snapshot.createdChannels.push(options);
        return Promise.resolve({ ok: true, channel: { id: channelId, name: options.name } });
    }

    setChannelTopic(options: slack.ConversationsSetTopicArguments): Promise<slack.ConversationsSetTopicResponse> {
        this.snapshot.setChannelTopics.push(options);
        return Promise.resolve({ ok: true });
    }

    inviteToChannel(options: slack.ConversationsInviteArguments): Promise<slack.ConversationsInviteResponse> {
        this.snapshot.invitesToChannels.push(options);
        return Promise.resolve({ ok: true });
    }

    archiveChannel(options: slack.ConversationsArchiveArguments): Promise<slack.ConversationsArchiveResponse> {
        this.snapshot.archivedChannels.push(options);
        return Promise.resolve({ ok: true });
    }

    sendMessage(options: slack.ChatPostMessageArguments): Promise<slack.ChatPostMessageResponse> {
        this.snapshot.sentMessages.push(options);
        return Promise.resolve({ ok: true, channel: channelId });
    }
}
