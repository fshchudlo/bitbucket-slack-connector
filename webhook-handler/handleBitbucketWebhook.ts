import {
    sendCompletionMessageAndCloseTheChannel,
    setChannelTopicAndInviteParticipants,
    sendMessageAboutAddedComment,
    updateChannelMembers,
    sendMessageAboutNewCommit,
    sendMessageAboutReviewerAction,
    sendMessageAboutDeletedComment, sendMessageAboutEditedComment
} from "./use-cases";
import { SlackAPIAdapter } from "./SlackAPIAdapter";
import { BitbucketGateway } from "./BitbucketGateway";
import { sendMessageAboutPRModification } from "./use-cases/sendMessageAboutPRModification";
import { BitbucketNotification } from "../typings";
import { buildChannelName } from "./slack-helpers";

export default async function handleBitbucketWebhook(payload: BitbucketNotification, slackAPI: SlackAPIAdapter, bitbucketGateway: BitbucketGateway, usePrivateChannels: boolean = true) {
    const eventKey = payload.eventKey;
    const channelInfo = await provisionPullRequestChannel(slackAPI, payload, usePrivateChannels);
    switch (eventKey) {
        case "pr:opened":
            await setChannelTopicAndInviteParticipants(payload, slackAPI, channelInfo);
            break;
        case "pr:modified":
            await sendMessageAboutPRModification(payload, slackAPI, channelInfo);
            break;
        case "pr:reviewer:updated":
            await updateChannelMembers(payload, slackAPI, channelInfo);
            break;
        case "pr:reviewer:unapproved":
        case "pr:reviewer:needs_work":
        case "pr:reviewer:approved":
            await sendMessageAboutReviewerAction(payload, slackAPI, channelInfo);
            break;
        case "pr:comment:added":
            await sendMessageAboutAddedComment(payload, slackAPI, channelInfo);
            break;
        case "pr:comment:edited":
            await sendMessageAboutEditedComment(payload, slackAPI, channelInfo);
            break;
        case "pr:comment:deleted":
            await sendMessageAboutDeletedComment(payload, slackAPI, channelInfo);
            break;
        case "pr:from_ref_updated":
            await sendMessageAboutNewCommit(payload, slackAPI, bitbucketGateway, channelInfo);
            break;
        case "pr:merged":
        case "pr:declined":
        case "pr:deleted":
            await sendCompletionMessageAndCloseTheChannel(payload, slackAPI, channelInfo);
            break;
        default:
            throw new Error(`"${eventKey}" event key is unknown.`);
    }
}

async function provisionPullRequestChannel(slackAPI: SlackAPIAdapter, payload: BitbucketNotification, usePrivateChannels: boolean) {
    return payload.eventKey == "pr:opened" ? await slackAPI.createChannel({
        name: buildChannelName(payload.pullRequest),
        isPrivate: usePrivateChannels
    }) : await slackAPI.findChannel(buildChannelName(payload.pullRequest), true);
}
