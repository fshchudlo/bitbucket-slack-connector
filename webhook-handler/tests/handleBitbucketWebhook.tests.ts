import TestSlackGateway from "./TestSlackGateway";
import TestPayloadBuilder from "./TestPayloadBuilder";
import handleBitbucketWebhook from "../handleBitbucketWebhook";
import { TestBitbucketGateway } from "./TestBitbucketGateway";
import { PullRequestBasicNotification } from "../../typings";

let testSlackGateway: TestSlackGateway = null;
let testBitbucketGateway: TestBitbucketGateway = null;
describe("handleBitbucketWebhook", () => {
    beforeEach(() => {
        testSlackGateway = new TestSlackGateway();
        testBitbucketGateway = new TestBitbucketGateway();
    });
    it("Should create channel, set topic and invite author and reviewers on PR opened", async () => {
        const payload = TestPayloadBuilder.pullRequestOpened();

        await handleBitbucketWebhook(payload, testSlackGateway, testBitbucketGateway);

        expect(testSlackGateway.snapshot).toMatchSnapshot();
    });

    it("Should send message on PR comment", async () => {
        const payload = TestPayloadBuilder.pullRequestCommentAdded();
        await handleBitbucketWebhook(payload, testSlackGateway, testBitbucketGateway);

        expect(testSlackGateway.snapshot).toMatchSnapshot();
    });

    it("Should send message on PR comment modification", async () => {
        const payload = TestPayloadBuilder.pullRequestCommentEdited();
        await handleBitbucketWebhook(payload, testSlackGateway, testBitbucketGateway);

        expect(testSlackGateway.snapshot).toMatchSnapshot();
    });

    it("Should send message on PR comment deletion", async () => {
        const payload = TestPayloadBuilder.pullRequestCommentDeleted();
        await handleBitbucketWebhook(payload, testSlackGateway, testBitbucketGateway);

        expect(testSlackGateway.snapshot).toMatchSnapshot();
    });

    it("Should send message on PR modification", async () => {
        const payload = TestPayloadBuilder.pullRequestModified();
        await handleBitbucketWebhook(payload, testSlackGateway, testBitbucketGateway);

        expect(testSlackGateway.snapshot).toMatchSnapshot();
    });

    it("Should send message on PR approval/unapproval/needs work", async () => {
        await handleBitbucketWebhook(TestPayloadBuilder.pullRequestNeedsWork(), testSlackGateway, testBitbucketGateway);
        await handleBitbucketWebhook(TestPayloadBuilder.pullRequestUnapproved(), testSlackGateway, testBitbucketGateway);
        await handleBitbucketWebhook(TestPayloadBuilder.pullRequestApproved(), testSlackGateway, testBitbucketGateway);
        expect(testSlackGateway.snapshot).toMatchSnapshot();
    });

    it("Should send message on PR commit", async () => {
        const payload = TestPayloadBuilder.pullRequestFromRefUpdated();
        await handleBitbucketWebhook(payload, testSlackGateway, testBitbucketGateway);

        expect(testSlackGateway.snapshot).toMatchSnapshot();
    });

    it("Should update channel participants on reviewers list update", async () => {
        const payload = TestPayloadBuilder.reviewersUpdated();
        await handleBitbucketWebhook(payload, testSlackGateway, testBitbucketGateway);

        expect(testSlackGateway.snapshot).toMatchSnapshot();
    });

    it("Should send completion message and close the channel on PR merge", async () => {
        const payload = TestPayloadBuilder.pullRequestMerged();
        await handleBitbucketWebhook(payload, testSlackGateway, testBitbucketGateway);

        expect(testSlackGateway.snapshot).toMatchSnapshot();
    });

    it("Should send completion message and close the channel on PR declining", async () => {
        const payload = TestPayloadBuilder.pullRequestDeclined();

        await handleBitbucketWebhook(payload, testSlackGateway, testBitbucketGateway);

        expect(testSlackGateway.snapshot).toMatchSnapshot();
    });

    it("Should send completion message and close the channel on PR deletion", async () => {
        const payload = TestPayloadBuilder.pullRequestDeleted();

        await handleBitbucketWebhook(payload, testSlackGateway, testBitbucketGateway);

        expect(testSlackGateway.snapshot).toMatchSnapshot();
    });

    it("Should throw Error on unknown action type", async () => {
        expect.assertions(1);

        const invalidPayload = ({ eventKey: "unknown action" } as unknown) as PullRequestBasicNotification;

        try {
            await handleBitbucketWebhook(invalidPayload, testSlackGateway, testBitbucketGateway);
        } catch (error) {
            expect((error as Error).message).toBe("\"unknown action\" event key is unknown.");
        }
    });
});
