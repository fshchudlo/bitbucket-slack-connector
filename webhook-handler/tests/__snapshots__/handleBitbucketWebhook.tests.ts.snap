// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleBitbucketWebhook Should create channel, set topic and invite author and reviewers on PR opened 1`] = `
{
  "archivedChannels": [],
  "createdChannels": [
    {
      "is_private": true,
      "name": "pr-test-proj-test-repository-1",
    },
  ],
  "invitesToChannels": [
    {
      "channel": "12345",
      "force": true,
      "users": "test.author@test.com,test.reviewer1@test.com",
    },
  ],
  "kicksFromChannels": [],
  "lookedUpUsers": [
    {
      "email": "test.author@test.com",
    },
    {
      "email": "test.reviewer1@test.com",
    },
  ],
  "sentMessages": [
    {
      "attachments": [
        {
          "color": "#0288D1",
          "text": "Quite a long and comprehensive description to get channel title longer than 250 symbols which is a maximum length for the Slack",
          "title": "The pull request was opened by Test Author.",
        },
        {
          "color": "#0288D1",
          "text": "Please <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview|review the PR>",
        },
      ],
      "channel": "12345",
      "icon_emoji": ":bitbucket:",
    },
  ],
  "setChannelTopics": [
    {
      "channel": "12345",
      "topic": ":git: Pull request: Quite a long and comprehensive title to get channel title longer than 250 symbols which is a maximum length for the Slack | :git-branch: To branch: \`TEST-PROJ/test-repository:master\`",
    },
  ],
}
`;

exports[`handleBitbucketWebhook Should send completion message and close the channel on PR declining 1`] = `
{
  "archivedChannels": [
    {
      "channel": "12345",
    },
  ],
  "createdChannels": [],
  "invitesToChannels": [],
  "kicksFromChannels": [],
  "lookedUpUsers": [],
  "sentMessages": [
    {
      "attachments": [
        {
          "color": "#9E34EB",
          "text": "The pull request was declined by Test Author.",
        },
      ],
      "channel": "pr-test-proj-test-repository-1",
      "icon_emoji": ":bitbucket:",
    },
  ],
  "setChannelTopics": [],
}
`;

exports[`handleBitbucketWebhook Should send completion message and close the channel on PR deletion 1`] = `
{
  "archivedChannels": [
    {
      "channel": "12345",
    },
  ],
  "createdChannels": [],
  "invitesToChannels": [],
  "kicksFromChannels": [],
  "lookedUpUsers": [],
  "sentMessages": [
    {
      "attachments": [
        {
          "color": "#9E34EB",
          "text": "The pull request was deleted by Test Author.",
        },
      ],
      "channel": "pr-test-proj-test-repository-1",
      "icon_emoji": ":bitbucket:",
    },
  ],
  "setChannelTopics": [],
}
`;

exports[`handleBitbucketWebhook Should send completion message and close the channel on PR merge 1`] = `
{
  "archivedChannels": [
    {
      "channel": "12345",
    },
  ],
  "createdChannels": [],
  "invitesToChannels": [],
  "kicksFromChannels": [],
  "lookedUpUsers": [],
  "sentMessages": [
    {
      "attachments": [
        {
          "color": "#4CAF50",
          "text": "The pull request was merged by Test Author. Well done, thank you all.",
        },
      ],
      "channel": "pr-test-proj-test-repository-1",
      "icon_emoji": ":bitbucket:",
    },
  ],
  "setChannelTopics": [],
}
`;

exports[`handleBitbucketWebhook Should send message on PR approval/unapproval/needs work 1`] = `
{
  "archivedChannels": [],
  "createdChannels": [],
  "invitesToChannels": [],
  "kicksFromChannels": [],
  "lookedUpUsers": [],
  "sentMessages": [
    {
      "attachments": [
        {
          "color": "#F6C342",
          "text": "Unapproved: Test Reviewer 1",
          "title": "Test Reviewer 1 requested changes for <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview|pull request>.",
        },
      ],
      "channel": "pr-test-proj-test-repository-1",
      "icon_emoji": ":bitbucket:",
    },
    {
      "attachments": [
        {
          "color": "#F6C342",
          "text": "Unapproved: Test Reviewer 1",
          "title": "Test Reviewer 1 unapproved <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview|pull request>.",
        },
      ],
      "channel": "pr-test-proj-test-repository-1",
      "icon_emoji": ":bitbucket:",
    },
    {
      "attachments": [
        {
          "color": "#4CAF50",
          "text": "Unapproved: Test Reviewer 1",
          "title": "Test Reviewer 1 approved <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview|pull request>.",
        },
      ],
      "channel": "pr-test-proj-test-repository-1",
      "icon_emoji": ":bitbucket:",
    },
  ],
  "setChannelTopics": [],
}
`;

exports[`handleBitbucketWebhook Should send message on PR comment 1`] = `
{
  "archivedChannels": [],
  "createdChannels": [],
  "invitesToChannels": [],
  "kicksFromChannels": [],
  "lookedUpUsers": [],
  "sentMessages": [
    {
      "attachments": [
        {
          "color": "#F6C342",
          "text": "> Test comment",
          "title": "Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|commented>:",
        },
      ],
      "channel": "pr-test-proj-test-repository-1",
      "icon_emoji": ":bitbucket:",
    },
  ],
  "setChannelTopics": [],
}
`;

exports[`handleBitbucketWebhook Should send message on PR comment deletion 1`] = `
{
  "archivedChannels": [],
  "createdChannels": [],
  "invitesToChannels": [],
  "kicksFromChannels": [],
  "lookedUpUsers": [],
  "sentMessages": [
    {
      "attachments": [
        {
          "color": "#0288D1",
          "text": "> Test comment",
          "title": "Test Reviewer 1 deleted comment:",
        },
      ],
      "channel": "pr-test-proj-test-repository-1",
      "icon_emoji": ":bitbucket:",
    },
  ],
  "setChannelTopics": [],
}
`;

exports[`handleBitbucketWebhook Should send message on PR comment modification 1`] = `
{
  "archivedChannels": [],
  "createdChannels": [],
  "invitesToChannels": [],
  "kicksFromChannels": [],
  "lookedUpUsers": [],
  "sentMessages": [
    {
      "attachments": [
        {
          "color": "#9E34EB",
          "text": "> Updated comment text",
          "title": "Test Reviewer 1 <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview?commentId=1|edited comment>:",
        },
      ],
      "channel": "pr-test-proj-test-repository-1",
      "icon_emoji": ":bitbucket:",
    },
  ],
  "setChannelTopics": [],
}
`;

exports[`handleBitbucketWebhook Should send message on PR commit 1`] = `
{
  "archivedChannels": [],
  "createdChannels": [],
  "invitesToChannels": [],
  "kicksFromChannels": [],
  "lookedUpUsers": [],
  "sentMessages": [
    {
      "attachments": [
        {
          "color": "#F6C342",
          "text": "> Test comment for TEST-PROJ, test-repository, from-ref-commit-hash

Please <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview|review the PR>.",
          "title": "A <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/commits/from-ref-commit-hash|new commit> was added to the pull request by Test Author.",
        },
      ],
      "channel": "pr-test-proj-test-repository-1",
      "icon_emoji": ":bitbucket:",
    },
  ],
  "setChannelTopics": [],
}
`;

exports[`handleBitbucketWebhook Should send message on PR modification 1`] = `
{
  "archivedChannels": [],
  "createdChannels": [],
  "invitesToChannels": [],
  "kicksFromChannels": [],
  "lookedUpUsers": [],
  "sentMessages": [
    {
      "attachments": [
        {
          "color": "#0288D1",
          "text": "Title is changed to:

> New pull request title

Description is changed to:

New pull request description

Target is changed to \`not-the-master\`

Please <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview|review the PR>",
          "title": "Test Author changed the pull request",
        },
      ],
      "channel": "pr-test-proj-test-repository-1",
      "icon_emoji": ":bitbucket:",
    },
  ],
  "setChannelTopics": [],
}
`;

exports[`handleBitbucketWebhook Should update channel participants on reviewers list update 1`] = `
{
  "archivedChannels": [],
  "createdChannels": [],
  "invitesToChannels": [
    {
      "channel": "12345",
      "force": true,
      "users": "test.reviewer2@test.com",
    },
  ],
  "kicksFromChannels": [
    {
      "channel": "12345",
      "user": "test.reviewer1@test.com",
    },
  ],
  "lookedUpUsers": [
    {
      "email": "test.reviewer2@test.com",
    },
    {
      "email": "test.reviewer1@test.com",
    },
  ],
  "sentMessages": [],
  "setChannelTopics": [],
}
`;
